pallets/stateful-storage/src/lib.rs

```diff
@@ -0,0 +1,777 @@
+		PageIdExceedsMaxAllowed,
+
+		/// Page size exceeds max allowable page size
+		PageExceedsMaxPageSizeBytes,
+
+		/// Invalid SchemaId or Schema not found
+		InvalidSchemaId,
+
+		/// Unsupported operation for schema
+		UnsupportedOperationForSchema,
+
+		/// Schema is not valid for storage model
+		SchemaPayloadLocationMismatch,
+
+		/// Invalid Message Source Account
+		InvalidMessageSourceAccount,
+
+		/// UnauthorizedDelegate
+		UnauthorizedDelegate,
+
+		/// Corrupted State
+		CorruptedState,
+
+		/// Invalid item action
+		InvalidItemAction,
+
+		/// Target page hash does not match current page hash
+		StalePageState,
+
+		/// Invalid Signature for payload
+		InvalidSignature,
+
+		/// The submitted proof has expired; the current block is less the expiration block
+		ProofHasExpired,
+
+		/// The submitted proof expiration block is too far in the future
+		ProofNotYetValid,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An event for when an itemized storage is updated
+		ItemizedPageUpdated {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// previous content hash before update
+			prev_content_hash: PageHash,
+			/// current content hash after update
+			curr_content_hash: PageHash,
+		},
+
+		/// An event for when an itemized storage is deleted
+		ItemizedPageDeleted {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// previous content hash before removal
+			prev_content_hash: PageHash,
+		},
+
+		/// An event for when an paginated storage is updated
+		PaginatedPageUpdated {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// id of updated page
+			page_id: PageId,
+			/// previous content hash before update
+			prev_content_hash: PageHash,
+			/// current content hash after update
+			curr_content_hash: PageHash,
+		},
+
+		/// An event for when an paginated storage is deleted
+		PaginatedPageDeleted {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// id of updated page
+			page_id: PageId,
+			/// previous content hash before removal
+			prev_content_hash: PageHash,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Applies the Add or Delete Actions on the requested Itemized page
+		///
+		/// # Events
+		/// * [`Event::ItemizedPageUpdated`]
+		/// * [`Event::ItemizedPageDeleted`]
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::apply_item_actions( actions.len() as u32 ,
+			actions.iter().fold(0, |acc, a| acc.saturating_add(match a {
+				ItemAction::Add { data } => data.len() as u32,
+				_ => 0,
+			})),
+		    0
+		))]
+		pub fn apply_item_actions(
```

Comment:
In the spirit of sql transactions were commit is called to save a transaction, perhaps, we use commit instead of apply.

Reply:
This is like a transaction where we are doing multiple operations. In SQL commit is what is used to save transactions.

Reply:
All extrinsic calls are transactions anyway if I remember correctly. I don't see any value in indicating the obvious.

Reply:
What could be more obvious is that you have a batch of operations users wants to perform. We can write this in a manner in which we discard the failed ones and continue with the rest of the operations. So it could be more evident that these actions are transactional. Recall that I had to ask you this question myself? I was not sure what actions would be applied. For this reason, it's not apparent. 

However, it's not a big deal; I only recommended this to align more with db operations. It's good as is.

Reply:
That is an issue that we seems like we should handle by adding more documentation.